### ISSUE TITLE
Deepcopy of bfloat16 array messes up bfloat16 definition.

### ISSUE THREAD

**Issue by frsong** [google/jax#4651]
After deepcopying a bfloat16 JAX array, operations on bfloat16 arrays lead to type errors. This seems to originate in dtype definitions becoming inconsistent.

<img width="1606" alt="Screenshot 2020-10-20 at 12 11 00" src="https://user-images.githubusercontent.com/12862386/96578713-8c842c00-12cd-11eb-8aa7-72caa60c14f6.png">

**Comment by girving** [comment]
This is delightful. :)

**Comment by hawkinsp** [comment]
I think two things are happening here:
a) `deepcopy.copy` of a `DeviceArray` produces a classic NumPy array. (This seems a bit surprising to me.)

b)

```
In [14]: x = jnp.ones((2,), jnp.bfloat16)

In [15]: z = np.ones((2,), jnp.bfloat16)

In [16]: x == z
...
TypeError: Invalid type promotion of bfloat16 and bfloat16
```

That's certainly a bug. (NumPy dtype extensions are tricky...)

**Comment by hawkinsp** [comment]
Actually I take it back. The repro I showed only fails after you have first done a deepcopy.

**Comment by hawkinsp** [comment]
It turns out our bfloat16 extension was not correctly initializing the hash value for the dtype descriptor.

I have a fix coming shortly; it will require building jaxlib from source or waiting for a new jaxlib release.

**Comment by mattjj** [comment]
@hawkinsp I think the deepcopy behavior is from [this definition of `__reduce__` on `DeviceArray`](https://github.com/google/jax/blob/537427aae26127df82faac4c82d6b4475d25f919/jax/interpreters/xla.py#L1148), since `deepcopy.copy` basically pickles and unpickles. We could define a `__deepcopy__` to avoid this behavior. WDYT?

**Comment by hawkinsp** [comment]
I think the behavior where `.copy()` and deepcopy return NumPy arrays is surprising and we should fix. Not in this issue, though!

**Comment by levskaya** [comment]
Yeah the deepcopy issue was impacting a user just yesterday whom I was helping, and I was completely flummoxed by why things were ending up back on host!

**Comment by hawkinsp** [comment]
This fix has been deployed in jaxlib 0.1.57. Hope that helps!

**Comment by frsong** [comment]
Thanks very much! And thanks for Jax!
